[RFC] Add runtime/neovim.desktop

teto opened the pull request:
Not quite sure where to put the icon ?

In 0f604e1da21c035db34f5f194a867829d58c55cc vim patch 904 was marked as NA but this commit reverts the decision via adding runtime/neovim.desktop.

Specifications available at:
https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#recognized-keys

Icons are stored system-wide in /usr/share/applications or user wide at
/usr/share/icons/hicolor/scalable/apps and can be overriden in ~/.local/share/icons

Likewise, the neovim.desktop file can be installed system wide or in
~/.local/share/applications/

To test without an installer:
$ xdg-desktop-menu install --novendor runtime/neovim.desktop
$ xdg-icon-resource install --novendor --mode user --size 64 contrib/neovim-icon.png

Once it is installed, you can test with gtk-launch if installed or
dmenu/rofi (drun mode)

teto committed:
Add runtime/neovim.desktop

In 0f604e1da21c035db34f5f194a867829d58c55cc vim patch 904 was marked as NA but this commit reverts the decision via adding runtime/neovim.desktop.

Specifications available at:
https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#recognized-keys

Icons are stored system-wide in /usr/share/applications or user wide at
/usr/share/icons/hicolor/scalable/apps and can be overriden in ~/.local/share/icons

Likewise, the neovim.desktop file can be installed system wide or in
~/.local/share/applications/

To test without an installer:
$ xdg-desktop-menu install --novendor runtime/neovim.desktop
$ xdg-icon-resource install --novendor --mode user --size 64 contrib/neovim-icon.png

Once it is installed, you can test with gtk-launch if installed or
dmenu/rofi (drun mode)

justinmk commented:
Does cmakelist.txt need a change? Or it's up to package maintainers to do something with this?

teto commented:
IMO it's up to package maintainers but I never packaged for a distribution so that's pure conjecture.

I could add a target to the makefile that would run:
$ xdg-desktop-menu install --novendor runtime/neovim.desktop
$ xdg-icon-resource install --novendor --mode user --size 64 contrib/neovim-icon.png
That's up to you.

Btw, is the icon location/choice ok ? I am not sure it's even used (at least not with my WM)?

justinmk commented on runtime/neovim.desktop:
```diff
@@ -0,0 +1,81 @@
+Terminal=true
+Type=Application
+Keywords=Text;editor;
+Icon=neovim-icon
```
Should be the same name as the executable, `Icon=nvim`. https://freedesktop.org/wiki/Howto_desktop_files/

Perhaps the file should be named `nvim.desktop` also.

Can you check `desktop-file-validate nvim.desktop` if you have that available?

justinmk commented:
> Btw, is the icon location/choice ok ? I am not sure it's even used (at least not with my WM)?

Vim has `runtime/icons/` , that's where ours can go. However they should be _installed_ to whatever the standard place is, via cmake. 

One way: https://techbase.kde.org/Development/Tutorials/Desktop_File#cmake_based_projects

Different suggestion, more complete but requires detecting some tools: https://cmake.org/pipermail/cmake/2010-September/039410.html

```
find_program(XDG-MIME_EXECUTABLE xdg-mime)
find_program(XDG-DESKTOP-MENU_EXECUTABLE xdg-desktop-menu)
execute_process(COMMAND ${XDG-DESKTOP-MENU_EXECUTABLE} install
--novendor nvim.desktop)
execute_process(COMMAND ${XDG-MIME_EXECUTABLE} default nvim.desktop
"application/x-nvim-item")
```

justinmk commented:
If that's too much trouble don't worry about it, just include `nvim.desktop` and `nvim.svg` at top level of `runtime/`.

jamessan commented:
> Eh... this is sounding very distro-specific, let's not worry about it

It's not distro specific at all.  It's defined by the [icon theme] FDO spec.  The [Installing Application Icons] section, in particular, should help.

[icon theme]: https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html
[Installing Application Icons]: https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html#install_icons

justinmk commented:
Ok, I wasn't sure if those deep `share/icons/hicolor/48x48` hierarchies were standardized.

teto committed:
cmake: add 'desktop-install' and 'icon-install'

... targets when possible. 'runtime' target will trigger them.

teto commented:
You can't install svg with xdg-icon-resource hence I resized it and converted the logo to png (https://bugs.launchpad.net/ubuntu/+source/xdg-utils/+bug/790449).

I ran desktop-file-validate and it was successful.

I've added a 'desktop' target that installs in user mode. I don't want to spend too much time on this as I see this mainly as a convenience for package developers and people.

If someone could confirm the icon works, my WM www.i3wm.org doesn't display icons.

justinmk commented on CMakeLists.txt:
```diff
@@ -461,6 +461,16 @@ add_subdirectory(config)
 
+
+# add 'desktop' target to install neovim.desktop
+if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
```
Should go in `runtime/CMakeLists.txt` and installed as part of the `runtime` target.

instead of checking system name, how about checking if `xdg-desktop-menu` is available?

```
find_program(XDG_MENU_PRG xdg-desktop-menu)
find_program(XDG_ICON_PRG xdg-icon-resource)

if(XDG_MENU_PRG)
...
endif()

if(XDG_ICON_PRG)
...
endif()

```

teto replied:
Like it and updated the code with this way.

justinmk commented:
I tested this after `apt-get install xdg-utils` on a non-desktop ubuntu machine and got this:

```
[2/6] cd /home/vagrant/neovim/build/runtime && xdg-desktop-menu install --novendor /home/vagrant/neovim/runtime/nvim.desktop               
FAILED: cd /home/vagrant/neovim/build/runtime && xdg-desktop-menu install --novendor /home/vagrant/neovim/runtime/nvim.desktop             
xdg-desktop-menu: No writable system menu directory found. 
```

We don't want `make install` to fail for this, so I removed the `add_dependencies()` lines. If there's a better way, let me know. But at least we have the targets and files available for package maintainers, and users that want them.

teto commented:
At the beginning I had the flag `--user mode ` but that would not have been consistent with a systemwide install so I removed it. There is no silver bullet so removing them was the wisest choice.

Thanks for merging, my first contribution \o/
